### DISCUSSION TITLE
Annotation for form-configuration

### DISCUSSION THREAD

**Discussion by SBolsec** [croz-ltd/nrich#122] [Ideas]
## Idea

Instead of having to define mappings between a form id and the corrresponding class through property files, an annotation could be added to the class which defines the form id.

### Current

```properties
nrich.form-configuration:
  default-converter-enabled: true
  form-configuration-mapping:
    user.create-form: net.croz.nrich.example.CreateUserRequest
    user.update-form: net.croz.nrich.example.UpdateUserRequest
```

### Proposed

```java
@Form("user.create-form")
public class CreateUserRequest {
   ...
}
```

## Open questions

- if both are used, the annotation and propery file, which would take precedence
- define a name for the annotation

**Comment by agrancaric** [comment]
This seems like a good idea. I would probably give precedence to annotation. For name maybe something like `@FormConfigurationHolder` or just `@FormConfiguration`.

**Reply by jzrilic** [reply]
If we agree to go with this, I think we should probably give precedence to the property file. It's a common way of overriding things everywhere and it's configurable by nature. It would be strange if you have a configuration in the property file (which is configurable) that is overridden by annotation (which is not configurable; you have to change your code and recompile it to apply the change).

**Reply by agrancaric** [reply]
Fine by me.

**Comment by jzrilic** [comment]
One thing I don't like about this annotation approach is that you will end with dependency on some third-party library (in this case nrich) in your api module which contains request classes. Consequently, the request class is no longer clean. For example, for this reason, I avoid using Swagger annotations in request classes for API documentation.

**Reply by agrancaric** [reply]
Yes but this is optional, so just another way of doing the same thing, if you don't want to use it you don't have to you can still do the same thing through properties.

**Reply by jzrilic** [reply]
That's true. In that case, maybe we can go with this. In the end, it's up to the consumer to decide which approach fits better for his particular case.

**Comment by pzrinscak** [comment]
I like this feature :).

**Comment by jzrilic** [comment]
We agreed to go with this. @SBolsec Can you please open the enhancement request? Any further discussion on this topic can be continued there.

**Reply by SBolsec** [reply]
I added the enhancement request #124
